daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala

```diff
@@ -0,0 +1,335 @@
+    trd: TermRuntimeData,
+    pstate: PState,
+    priorState: PState.Mark,
+    maybeStartState: Maybe[PState.Mark],
+    ais: GoArrayIndexStatus): ParseAttemptStatus
+
+  final protected def tryParseDetectMarkLeaks(
+    parser: SequenceChildParser,
+    pstate: PState,
+    maybeStartState: Maybe[PState.Mark],
+    ais: GoArrayIndexStatus): ParseAttemptStatus = {
+
+    var markLeakCausedByException = false
+    var priorState: PState.Mark = null
+    val result =
+      try {
+        priorState = pstate.mark("OrderedSeparatedSequence_beforeSeparator")
+        val res = parseOne(parser, parser.trd, pstate, priorState, maybeStartState, ais)
+        priorState = null
+        res
+      } catch {
+        // Similar try/catch/finally logic for returning marks is also used in
+        // the AltCompParser and RepUnboundedParser. The logic isn't
+        // easily factored out so it is duplicated. Changes made here should also
+        // be made there. Only these parsers deal with taking marks, so this logic
+        // should not be needed elsewhere.
+        case t: Throwable => {
+          if (priorState != null) {
+            markLeakCausedByException = true
+            if (!t.isInstanceOf[SchemaDefinitionDiagnosticBase] && !t.isInstanceOf[UnsuppressableException]) {
+              val stackTrace = new StringWriter()
+              t.printStackTrace(new PrintWriter(stackTrace))
+              Assert.invariantFailed("Exception thrown with mark not returned: " + t + "\nStackTrace:\n" + stackTrace)
+            }
+          }
+          throw t
+        }
+      } finally {
+        var markLeak = false;
+        if (priorState != null) {
+          pstate.discard(priorState)
+          markLeak = true;
+        }
+        if (markLeak && !markLeakCausedByException) {
+          // likely a logic bug, throw assertion
+          Assert.invariantFailed("mark not returned, likely a logic bug")
+        }
+      } // end try/catch/finally
+    result
+  }
+
+  protected def failedChild(pstate: PState, priorState: PState.Mark): ParseAttemptStatus = {
+    val cause = pstate.processorStatus.asInstanceOf[Failure].cause
+    PE(pstate, "Failed to parse sequence child. Cause: %s.", cause)
+    ParseAttemptStatus.FailedSpeculativeParse
+  }
+
+  protected def processFailedChildParseResults(
+    pstate: PState,
+    priorState: PState.Mark,
+    maybeStartState: Maybe[PState.Mark]): ParseAttemptStatus = {
+    val isFixedOccurs = maybeStartState.isEmpty
+    val isVariableOccurs = !isFixedOccurs
+    //
+    // we failed to parse a child
+    //
+    // failedChild(pstate, priorState) // discards prior state
+    if (isFixedOccurs) {
+      //
+      // in fixed occurs, there are no points of uncertainty for the
+      // individual elements, so a failure is a failure of the whole loop.
+      // And any discriminator that has been set is the discriminator
+      // of some surrounding scope.
+      //
+      ParseAttemptStatus.FailedEntireArray
```

Comment:
Potential mark leak here?

Reply:
Yup. There *were* lots of these. I wasn't even trying to track them all down.